packages/messaging/lib/windows.js

```diff
@@ -32,32 +37,130 @@ export class WindowsMessagingTransport {
    * @param {{signal?: AbortSignal}} opts
    * @return {Promise<any>}
    */
-  // @ts-ignore
-  // eslint-disable-next-line @typescript-eslint/no-unused-vars
-  request(name, data = {}, opts = {}) {
-    return windowsTransport(this.config, name, data, opts).withResponse(name + 'Response')
+  request(msg, opts = {}) {
+
+    // convert the message to window-specific naming
+    const outgoing = WindowsRequestMessage.fromRequest(msg);
+
+    // send the message
+    this.config.methods.postMessage(outgoing)
+
+    // compare incoming messages against the `msg.id`
+    const comparator = (eventData) => {
+      return eventData.featureName === msg.featureName
+          && eventData.context === msg.context
+          && eventData.id === msg.id
+    }
+
+    // now wait for a matching message
+    return new Promise((resolve, reject) => {
+      try {
+        this._subscribe(comparator, opts, (value, unsubscribe) => {
+          unsubscribe();
+          if ('result' in value) {
+            resolve(value['result']);
+          } else if ('error' in value) {
+            // @ts-expect-error
+            reject(new Error(value.error.message || 'unknown error'))
+          } else {
+            console.warn('unknown response type', value);
+            reject(new Error('unknown response'))
+          }
+        })
+      } catch (e) {
+        reject(e)
+      }
+    })
+  }
+  /**
+   * @param {import("../index.js").Subscription} msg
+   * @param {(value: unknown | undefined) => void} callback
+   */
+  subscribe(msg, callback) {
+
+    // compare incoming messages against the `msg.subscriptionName`
+    const comparator = (eventData) => {
+      return eventData.featureName === msg.featureName
+          && eventData.context === msg.context
+          && eventData.subscriptionName === msg.subscriptionName
+    }
+
+    // only forward the 'params' from a SubscriptionEvent
+    const cb = (eventData) => {
+      return callback(eventData['params']);
+    }
+
+    // now listen for matching incoming messages.
+    return this._subscribe(comparator, {}, cb)
   }
 
   /**
-   * @param {string} name
-   * @param {(value: unknown) => void} callback
+   * @typedef {import("../index.js").MessageResponse | import("../index.js").SubscriptionEvent} Incoming
+   * @param {(eventData: any) => boolean} comparator
+   * @param {{signal?: AbortSignal}} options
+   * @param {(value: Incoming, unsubscribe: (()=>void)) => void} callback
+   * @internal
    */
-  subscribe(name, callback) {
-    return subscribe(this.config, name, {}, callback)
+  _subscribe(comparator, options, callback) {
+    // if already aborted, reject immediately
+    if (options?.signal?.aborted) {
+      throw new DOMException('Aborted', 'AbortError')
+    }
+    /** @type {(()=>void) | undefined} */
+    let teardown
+
+    /**
+     * @param {MessageEvent} event
+     */
+    const idHandler = (event) => {
+      if (this.messagingContext.env === "production") {
```

Comment:
We can ignore this also, just I've seen these general flags of prod/dev become a little unmaintainable to manage the differences rather than tiny specific items that can be turned off. (Both move the complexity somewhere else, so it's a case-by-case thing).

Reply:
yeah it's tough to know what's 'best' here - I think ensuring that 'npm run build' without the explicit flag is *always* production is importnt

Reply:
eg: you have to opt-in to 'development' mode